op-challenger/game/keccak/merkle/tree.go

```diff
@@ -0,0 +1,193 @@
+type BinaryMerkleTree struct {
+	Root      *merkleNode
+	LeafCount int
+}
+
+func NewBinaryMerkleTree() *BinaryMerkleTree {
+	return &BinaryMerkleTree{
+		Root:      &merkleNode{Label: common.Hash{}, GIndex: big.NewInt(0)},
+		LeafCount: 0,
+	}
+}
+
+// RootHash returns the root hash of the binary merkle tree.
+func (m *BinaryMerkleTree) RootHash() (rootHash common.Hash) {
+	if m.Root.Label == (common.Hash{}) {
+		for height := 0; height < BinaryMerkleTreeDepth; height++ {
+			rootHash = crypto.Keccak256Hash(rootHash[:], zeroHashes[height][:])
+		}
+		m.Root.Label = rootHash
+	}
+	return m.Root.Label
+}
+
+// walkDownToMaxLeaf walks down the tree to the max leaf node.
+func (m *BinaryMerkleTree) walkDownToLeafCount(subtreeLeafCount int) *merkleNode {
+	maxSubtreeLeafCount := MaxLeafCount + 1
+	levelNode := m.Root
+	for height := 0; height < BinaryMerkleTreeDepth; height++ {
+		if subtreeLeafCount*2 <= maxSubtreeLeafCount {
+			if levelNode.Left == nil {
+				levelNode.Left = &merkleNode{
+					Label:  zeroHashes[height],
+					GIndex: new(big.Int).Add(big.NewInt(1), new(big.Int).Mul(levelNode.GIndex, big.NewInt(2))),
+					Parent: levelNode,
+				}
+			}
+			levelNode = levelNode.Left
+		} else {
+			if levelNode.Right == nil {
+				levelNode.Right = &merkleNode{
+					Label:  zeroHashes[height],
+					GIndex: new(big.Int).Add(big.NewInt(2), new(big.Int).Mul(levelNode.GIndex, big.NewInt(2))),
+					Parent: levelNode,
+				}
+			}
+			levelNode = levelNode.Right
+			subtreeLeafCount -= maxSubtreeLeafCount / 2
+		}
+		maxSubtreeLeafCount /= 2
+	}
+	return levelNode
+}
+
+// AddLeaf adds a leaf to the binary merkle tree.
+func (m *BinaryMerkleTree) AddLeaf(leaf types.Leaf) {
+	// Walk down to the new max leaf node.
+	m.LeafCount += 1
+	levelNode := m.walkDownToLeafCount(m.LeafCount)
+
+	// Set the leaf node data.
+	levelNode.Label = leaf.Hash()
+
+	// Walk back up the tree, updating the hashes with its sibling hash.
+	for height := 0; height < BinaryMerkleTreeDepth; height++ {
+		if levelNode.Parent.IsLeftChild(levelNode) {
+			if levelNode.Parent.Right == nil {
+				levelNode.Parent.Right = &merkleNode{
+					Label:  zeroHashes[height],
+					GIndex: new(big.Int).Add(levelNode.GIndex, big.NewInt(1)),
+					Parent: levelNode.Parent,
+				}
+			}
+			levelNode.Parent.Label = crypto.Keccak256Hash(levelNode.Label[:], levelNode.Parent.Right.Label[:])
+		} else {
+			if levelNode.Parent.Left == nil {
+				levelNode.Parent.Left = &merkleNode{
+					Label:  zeroHashes[height],
+					GIndex: new(big.Int).Sub(levelNode.GIndex, big.NewInt(1)),
+					Parent: levelNode.Parent,
+				}
+			}
+			levelNode.Parent.Label = crypto.Keccak256Hash(levelNode.Parent.Left.Label[:], levelNode.Label[:])
+		}
+		levelNode = levelNode.Parent
+	}
+}
+
+// ProofAtIndex returns a merkle proof at the given leaf node index.
+func (m *BinaryMerkleTree) ProofAtIndex(index uint64) (proof [BinaryMerkleTreeDepth]common.Hash, err error) {
+	if index >= (1<<BinaryMerkleTreeDepth)-1 {
```

Comment:
```suggestion
	if index >= MaxLeafCount {
```

or is there something funky going on with operator precedence?